AMP Sharing Support - Include all services

MattGeri opened the pull request:
This code adds support for all the different sharing services that Jetpack makes available. It also include the ability to print the AMP page.

As per the discussion in [Issue 15399](https://github.com/Automattic/jetpack/issues/15399), We've opted to use the existing font based icons and override the default AMP styles. This gives us a consistent look between AMP and non-AMP modes.

Fixes #15399 
Fixes #14926 

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* This enhances an existing feature.
* P2 shortlink: https://wp.me/pbIAo1-w

#### Testing instructions:
1. Ensure the [AMP plugin](https://wordpress.org/plugins/amp/) is active
2. In AMP > Settings, select Transitional mode
3. Enable sharing in Jetpack > Settings > Sharing or Settings > Sharing
4. Visit a post on the front end and check that social icons are enabled
5. Add `?amp` to the end URL and make sure the AMP icons display as per step 4
6. Expected: Sharing icons look the same on the AMP and non-AMP versions of the front end website.

Non-AMP:
<img width="1280" alt="Screen Shot 2020-04-20 at 5 08 01 PM" src="https://user-images.githubusercontent.com/49400/79767435-a0080780-8329-11ea-9edb-4031ce43c1c5.png">

AMP:
<img width="1280" alt="Screen Shot 2020-04-20 at 5 08 08 PM" src="https://user-images.githubusercontent.com/49400/79767455-a5fde880-8329-11ea-89eb-4783e4d7c64a.png">

jeherve commented on 3rd-party/class.jetpack-amp-support.php:
```diff
@@ -361,28 +364,84 @@ public static function render_sharing_html( $markup, $sharing_enabled ) {
-				$sharing_links[] = "<!-- not supported: $id -->";
+			$args = array();
+			$skip = false;
+			switch ( $service->get_id() ) {
```
Part of me wonders if at this point, we could include this in each service's class (e.g. `Share_Facebook` [#](https://github.com/Automattic/jetpack/blob/3699f34bb90a2b0a2af221042dacacf777238d2b/modules/sharedaddy/sharing-sources.php#L955)) as a new method, to avoid duplicating things and make it clear what services are supported by seeing which ones add the method to their class. What do you think of such an approach?

MattGeri replied:
I actually had a similar thought pattern, but then when I looked at the code further, I noticed that all AMP functionality is abstracted away in to this single file/class. I figured that was a design decision and decided to leave anything that deals with AMP in to this file.

I originally had the AMP CSS being enqueued inside the Sharedaddy class too, but moved it all to this file.

I'm happy to move AMP related functionality in to the modules, but it would break what I believe to be an intentional design of having it abstracted to a single place. Would love to hear your thoughts with regards to this?

jeherve replied:
Yes, it was a conscious decision when we first worked on AMP compatibility. I think it's worth revisiting now that we will support more sharing services; at this point AMP support for the sharing module has become significant enough that we could include it in the sharing file, as long as we are clear on what the methods are used for in there.

What do you think?

MattGeri replied:
Makes sense. I'd propose that we include a method which renders the AMP markup like `get_display` does in the parent class. This can return false if AMP is not supported for a given module or if it needs customizations (as per the rules here), we can override the method in the individual service. I'll work on this and send a commit to the PR to get your thoughts.

MattGeri replied:
@jeherve Implemented in: https://github.com/Automattic/jetpack/pull/15497/commits/d27daaecc836373885a4e9825865499c8de1d616 Would appreciate your thoughts

jeherve commented on 3rd-party/class.jetpack-amp-support.php:
```diff
@@ -415,6 +474,16 @@ public static function amp_disable_sharedaddy_css( $enqueue ) {
+	 */
+	public static function amp_enqueue_sharing_css() {
+		if ( self::is_amp_request() ) {
+			wp_enqueue_style( 'sharedaddy-amp', plugin_dir_url( dirname( __FILE__ ) ) . 'modules/sharedaddy/amp-sharing.css', array(), JETPACK__VERSION );
+			wp_enqueue_style( 'social-logos' );
```
If the new stylesheet depends on `social-logos`, and since that stylesheet is already registered [here](https://github.com/Automattic/jetpack/blob/b96f66a73398cd79dd460127fa07855ee4b85fcf/_inc/social-logos.php#L9-L12), should we just list it as a dependency of the new stylesheet?

MattGeri replied:
Good catch!

jeherve commented on 3rd-party/class.jetpack-amp-support.php:
```diff
@@ -31,6 +31,7 @@ public static function init() {
 		// Sharing.
 		add_filter( 'jetpack_sharing_display_markup', array( 'Jetpack_AMP_Support', 'render_sharing_html' ), 10, 2 );
 		add_filter( 'sharing_enqueue_scripts', array( 'Jetpack_AMP_Support', 'amp_disable_sharedaddy_css' ) );
+		add_action( 'wp_head', array( 'Jetpack_AMP_Support', 'amp_enqueue_sharing_css' ) );
```
Could we enqueue this into `wp_enqueue_scripts` instead?

MattGeri replied:
Yep sure, will update shortly!

jeherve commented on modules/sharedaddy/amp-sharing.css:
```diff
@@ -0,0 +1,110 @@
+amp-social-share {
+	border-radius: 50%;
+	color: #fff !important;
+    -webkit-border-radius: 50%;
```
I'd recommend using tabs for indenting in the whole file, so no one inadvertently changes all spaces into tabs when they next edit this file in their editor.

MattGeri replied:
Sorry, this was a bad c/p from the non-AMP stylesheet. Will push fix shortly

MattGeri commented:
Thanks for the feedback @jeherve 

I've addressed the issues in a new commit.

I've also added a fix for #12639. The problem here was that the regex that should have replaced the non-AMP sharing markup was cutting out too early when there was a closing `div` inside a list item. This was the case for "Official Icons" which have `div`s inside the list items.

MattGeri commented:
@jeherve This has been updated as per your suggestions in the inline comments. One additional thing, Code Climate is failing based on there being too many methods in Sharing Source now that we've moved the AMP rendering. Is this a blocker?

jeherve commented:
> Code Climate is failing based on there being too many methods in Sharing Source now that we've moved the AMP rendering. Is this a blocker?

No worries there, absolutely not a blocker. 👍

MattGeri commented:
> I seem to be missing the CSS for some reason. Does it still work well for you?
> 
> <img alt="screenshot 2020-04-24 at 15 54 07" width="1344" src="https://user-images.githubusercontent.com/426388/80220681-7eb65c80-8644-11ea-9a65-3dba9d81ab1b.png">

Strange! Yea, I just tested again and it works as expected for me:

<img width="520" alt="Screen Shot 2020-04-24 at 5 27 59 PM" src="https://user-images.githubusercontent.com/49400/80229708-1b7ef700-8651-11ea-9b89-740426251c92.png">

I'll run some tests on other environments and report back so see if it works on those as well

MattGeri commented:
@jeherve I've tested in a few environments and I cannot replicate that issue you're experiencing. Could you try test again?

I've additionally applied the changes to my personal blog and it works as expected. See here:

https://mattgeri.com/blog/2019-year-in-review/?amp

<img width="1177" alt="Screen Shot 2020-04-30 at 3 45 04 PM" src="https://user-images.githubusercontent.com/49400/80717794-933d9d80-8af9-11ea-93bf-051b0c0b2af1.png">

MattGeri commented on modules/sharedaddy/amp-sharing.css:
```diff
@@ -0,0 +1,110 @@
+
+amp-social-share[type='print']::before {
+	content: '\f469';
+	color: #656565 !important;
```
Unfortunately not, this is the output without it:

<img width="255" alt="Screen Shot 2020-04-30 at 3 48 33 PM" src="https://user-images.githubusercontent.com/49400/80718196-21198880-8afa-11ea-962c-7019b3a9c179.png">

westonruter replied:
Is it getting overridden by an AMP style or some other style in Jetpack? What selector is overriding it? An alternative to `!important` would be a more-specific selector.

MattGeri replied:
@westonruter @jeherve The style rules here were copied from the current non-AMP stylesheet for consistency between the two share bars on AMP and non-AMP. The original stylesheet had the `!important` rules, i.e. for reddit: https://github.com/Automattic/jetpack/blob/master/modules/sharedaddy/sharing.css#L642

I _believe_ they would have been added intentionally to avoid theme conflicts on the styles. However,  I'm not apposed to removing them for the AMP stylesheet. Here's a commit that removes the parent level important rule and the rules for reddit and print. https://github.com/Automattic/jetpack/pull/15497/commits/2a1ffca14212c663b36647f2ad80dac94a23997f

jeherve commented:
I played with this a bit, and I seem to run into issues when using Reader mode only. Things look good in transitional and standard modes.

MattGeri commented on 3rd-party/class.jetpack-amp-support.php:
```diff
@@ -409,6 +415,14 @@ public static function amp_enqueue_sharing_css() {
+	 * For the AMP Reader mode template, include styles that we need.
+	 */
+	public static function amp_reader_sharing_css() {
+		echo file_get_contents( plugin_dir_path( dirname( __FILE__ ) ) . '_inc/social-logos/social-logos.css' ); // phpcs:ignore
```
@jeherve I see this method of echoing CSS from a file for Reader mode is already used for the Related Posts module. I couldn't think of a better way to do it, or a way to escape CSS, so I added the phpcs:ignore rule.

jeherve replied:
This is a bit problematic, not because of the echoing, but because those 2 files (and their minified versions, which we may want to prefer here) don't live in that same location in the WordPress.com environment.

We could have a check for WordPress.com in this function and echo different files depending on the environment. Unless there is a way we can rely on the styles' enqueue names instead and inject them to the Reader mode's page that way?

MattGeri replied:
@jeherve If we went the route of enqueued assets, we'd need to do something like a `wp_remote_get` to fetch the styles.

Quick question, would using `JETPACK__PLUGIN_FILE` instead of `dirname( __FILE__ )` solve this problem?

jeherve replied:
`JETPACK__PLUGIN_FILE` won't do the trick here I'm afraid, as the file's relative locations are different on the 2 platforms.

I've just created #15762 to help a bit, alongside an existing constant we have for the Sharing feature, `WP_SHARING_PLUGIN_DIR`. Once that's merged, we should be able to do something like this:

```diff
diff --git a/3rd-party/class.jetpack-amp-support.php b/3rd-party/class.jetpack-amp-support.php
index a9de75e86..9de0c9ac4 100644
--- a/3rd-party/class.jetpack-amp-support.php
+++ b/3rd-party/class.jetpack-amp-support.php
@@ -419,8 +419,20 @@ class Jetpack_AMP_Support {
 	 * For the AMP Reader mode template, include styles that we need.
 	 */
 	public static function amp_reader_sharing_css() {
-		echo file_get_contents( plugin_dir_path( dirname( __FILE__ ) ) . '_inc/social-logos/social-logos.css' ); // phpcs:ignore
-		echo file_get_contents( plugin_dir_path( dirname( __FILE__ ) ) . 'modules/sharedaddy/amp-sharing.css' ); // phpcs:ignore
+		/*
+		 * We'll need to output the full contents of the 2 files
+		 * in the head on AMP views. We can't rely on regular enqueues here.
+		 *
+		 * phpcs:disable WordPress.WP.AlternativeFunctions.file_get_contents_file_get_contents
+		 * phpcs:disable WordPress.Security.EscapeOutput.OutputNotEscaped
+		 */
+		echo file_get_contents( JETPACK_SOCIAL_LOGOS_DIR . '_inc/social-logos/social-logos.css' );
+		echo file_get_contents( WP_SHARING_PLUGIN_DIR . 'modules/sharedaddy/amp-sharing.css' );
+
+		/*
+		 * phpcs:enable WordPress.WP.AlternativeFunctions.file_get_contents_file_get_contents
+		 * phpcs:enable WordPress.Security.EscapeOutput.OutputNotEscaped
+		 */
 	}
 
 	/**

```

jeherve replied:
I've now merged the other PR, you should be good to go with this!

MattGeri commented:
Really good catch @jeherve. My latest change includes the styles for the Reader mode template.

jeherve commented on modules/sharedaddy/sharing-sources.php:
```diff
@@ -395,6 +431,15 @@ abstract class Deprecated_Sharing_Source extends Sharing_Source {
+	 *
+	 * @param \WP_Post $post The current post being viewed.
+	 */
+	final public function get_amp_display( $post ) { // phpcs:ignore
```
Here and below, could you add the specific rule we're ignoring, for clarity?

```suggestion
	final public function get_amp_display( $post ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable
```

MattGeri replied:
Sure! Pushing this change now

westonruter commented:
I just tested this in the Lovecraft theme, and I'm seeing an issue where in AMP the last sharing button is shifted down:

![image](https://user-images.githubusercontent.com/134745/82271678-5103c100-992d-11ea-90f2-8167cc38c60a.png)

I'm not seeing this in the non-AMP version however:

![image](https://user-images.githubusercontent.com/134745/82271702-60830a00-992d-11ea-8364-ed28aee0368a.png)

MattGeri commented:
@jeherve Thanks for providing the const paths. I've pushed a changed as you requested to use those in the loading of the assets in to AMP

MattGeri commented:
> I just tested this in the [Lovecraft](https://wordpress.org/themes/lovecraft/) theme, and I'm seeing an issue where in AMP the last sharing button is shifted down:
> 
> ![image](https://user-images.githubusercontent.com/134745/82271678-5103c100-992d-11ea-90f2-8167cc38c60a.png)
> 
> I'm not seeing this in the non-AMP version however:
> 
> ![image](https://user-images.githubusercontent.com/134745/82271702-60830a00-992d-11ea-8364-ed28aee0368a.png)

@westonruter I've looked in to this. The reason is that the Lovecraft theme has a rule that the last item in the post content div should have no bottom margin.

`.post-content *:last-child {
    margin-bottom: 0;
}`

In the case of AMP, when social sharing is injected to the bottom of the article, it becomes the last item. The reason why it's not showing up on desktop, is that the social share bar on desktop has an empty `li` as the last element. See here:

<img width="1129" alt="Screenshot 2020-05-21 at 20 16 11" src="https://user-images.githubusercontent.com/49400/82592273-eeb2f680-9ba0-11ea-9af0-0b14bf1fe9e2.png">

We could look to add an empty element to AMP too, but I'm not sure we should be designing our solution around a CSS rule in a specific theme. The other option is the override it with the `!important` rule, but we've looked to eliminate that above.

Open to other suggestions that you may have here.

westonruter commented:
> I've looked in to this. The reason is that the Lovecraft theme has a rule that the last item in the post content div should have no bottom margin.
> 
> `.post-content *:last-child { margin-bottom: 0; }`

@MattGeri I bet this is actually a bug in Lovecraft, then. I bet that @andersnoren actually intended this CSS instead:

```css
.post-content > *:last-child { margin-bottom: 0; }
```

So as to remove the bottom margin on the last element in the post content. This makes more sense than to remove all bottom margin from all last-child elements in the entire tree.

That being the case, I'd say there's nothing that Jetpack needs to do here.

andersnoren commented:
@MattGeri @westonruter Correct! I’m working on the 2.0 version of Lovecraft now, and this will be changed as part of that.

westonruter commented:
Not sure if I'm testing with an older build, but I'm also seeing some missing styles in Reader mode:

![image](https://user-images.githubusercontent.com/134745/83467812-d6a86600-a42f-11ea-806d-c52827255375.png)

Non-AMP shows:

![image](https://user-images.githubusercontent.com/134745/83467840-ea53cc80-a42f-11ea-8229-9959f4114fc3.png)

ashicus commented:
@westonruter Those missing Reader mode styles should be fixed now.

rickalee commented:
@jeherve What is best way to remove amp_reader_sharing_css from Reader mode? This is breaking our design and default CSS. remove_action() doesn't seem to be working. Could we get a filter to prevent this output?

jeherve commented:
> What is best way to remove amp_reader_sharing_css from Reader mode? This is breaking our design and default CSS. remove_action() doesn't seem to be working.

@rickalee Would something like this work?

```php
function jeherve_rm_amp_sharing() {
	if ( class_exists( 'Jetpack_AMP_Support' ) ) {
		remove_action( 'amp_post_template_css', array( 'Jetpack_AMP_Support', 'amp_reader_sharing_css' ) );
	}
}
add_action( 'init', 'jeherve_rm_amp_sharing' );
```

rickalee commented:
Thanks @jeherve I had it on wrong hook. Worked perfectly.

westonruter commented on 3rd-party/class.jetpack-amp-support.php:
```diff
@@ -420,6 +411,35 @@ public static function amp_disable_sharedaddy_css( $enqueue ) {
+		 * phpcs:disable WordPress.WP.AlternativeFunctions.file_get_contents_file_get_contents
+		 * phpcs:disable WordPress.Security.EscapeOutput.OutputNotEscaped
+		 */
+		echo file_get_contents( JETPACK_SOCIAL_LOGOS_DIR . 'social-logos.css' );
```
There's a bug on this line which I've opened a fix for: https://github.com/Automattic/jetpack/pull/19503
